web3/contract.py

```diff
@@ -893,11 +1092,271 @@ def _set_function_info(self) -> None:
+                abi=wallet_contract_definition["abi"]
+            )
+
+            # Not a real contract address
+            contract = ContractFactory("0x2f70d3d26829e412A602E83FE8EeBF80255AEeA5")
+
+            # Read "owner" public variable
+            addr = contract.functions.owner().call()
+
+        :param transaction: Dictionary of transaction info for web3 interface
+        :return: ``Caller`` object that has contract public functions
+            and variables exposed as Python methods
+        """
+        call_transaction = self._get_call_txparams(transaction)
+
+        block_id = parse_block_identifier(self.w3, block_identifier)
+
+        return call_contract_function(self.w3,
+                                      self.address,
+                                      self._return_data_normalizers,
+                                      self.function_identifier,
+                                      call_transaction,
+                                      block_id,
+                                      self.contract_abi,
+                                      self.abi,
+                                      state_override,
+                                      *self.args,
+                                      **self.kwargs
+                                      )
+
+    @classmethod
+    def factory(cls, class_name: str, **kwargs: Any) -> 'ContractFunction':
+        return PropertyCheckingFactory(class_name, (cls,), kwargs)(kwargs.get('abi'))
+
+    def transact(self, transaction: Optional[TxParams] = None) -> HexBytes:
+        setup_transaction = self._transact(transaction)
+        return transact_with_contract_function(
+            self.address,
+            self.w3,
+            self.function_identifier,
+            setup_transaction,
+            self.contract_abi,
+            self.abi,
+            *self.args,
+            **self.kwargs
+        )
+
+    def estimate_gas(
+        self, transaction: Optional[TxParams] = None,
+        block_identifier: Optional[BlockIdentifier] = None
+    ) -> int:
+        setup_transaction = self._estimate_gas(transaction)
+        return estimate_gas_for_function(
+            self.address,
+            self.w3,
+            self.function_identifier,
+            setup_transaction,
+            self.contract_abi,
+            self.abi,
+            block_identifier,
+            *self.args,
+            **self.kwargs
+        )
+
+    def build_transaction(self, transaction: Optional[TxParams] = None) -> TxParams:
+
+        built_transaction = self._build_transaction(transaction)
+        return build_transaction_for_function(
+            self.address,
+            self.w3,
+            self.function_identifier,
+            built_transaction,
+            self.contract_abi,
+            self.abi,
+            *self.args,
+            **self.kwargs
+        )
+
+
+class AsyncContractFunction(BaseContractFunction):
+
+    def __call__(self,
```

Comment:
Can we combine this `__call__` and the one from ContractFunction [here](https://github.com/ethereum/web3.py/pull/2441/files#diff-80614febadb93f268dcfddd91641e79f9bf98b5469b84ab1915edd1534970bd7R1231) and pull them up into the Base class? or are they fundamentally different?

Reply:
The reason that I copied those down to the inherited classes is that mypy does not like them in the base class. Here is an example of what mypy says when they are moved to the base class with the below signature. Most of those errors are due to the call() now not returning the right type which is weird because I moved __call__ which should not effect call() as I understand it.

Also, I went back and forth on some of these methods being in the base class because I tried not to make the base class so aware of the inherited classes where I could. Like in this example the base class would have to know about the two inherited classes because of the return typing. The flip side of that argument is, "is it really worth the complexity and extra code just to not make the base class aware of the inherited classes"

thoughts??

```
    def __call__(self,
                 *args: Any,
                 **kwargs: Any) -> Union['AsyncContractFunction', 'ContractFunction']:
```

```
ens/main.py:203: error: Incompatible return value type (got "Union[Coroutine[Any, Any, HexBytes], HexBytes]", expected "HexBytes")
ens/main.py:339: error: Incompatible return value type (got "Union[Coroutine[Any, Any, HexBytes], HexBytes]", expected "HexBytes")
ens/main.py:467: error: Incompatible return value type (got "Union[Coroutine[Any, Any, HexBytes], HexBytes]", expected "HexBytes")
web3/pm.py:253: error: Argument 1 to "wait_for_transaction_receipt" of "Eth" has incompatible type "Union[Coroutine[Any, Any, HexBytes], HexBytes]"; expected "Union[Hash32, HexBytes, HexStr]"
web3/pm.py:258: error: Incompatible return value type (got "Union[Coroutine[Any, Any, Any], Any]", expected "str")
web3/pm.py:265: error: "typing.Coroutine[Any, Any, Any]" object is not iterable
web3/pm.py:275: error: Incompatible return value type (got "Union[Coroutine[Any, Any, Any], Any]", expected "bytes")
web3/pm.py:282: error: "typing.Coroutine[Any, Any, Any]" object is not iterable
web3/pm.py:293: error: "typing.Coroutine[Any, Any, Any]" object is not iterable
web3/pm.py:297: error: Incompatible return value type (got "Union[Coroutine[Any, Any, Any], Any]", expected "bytes")
web3/pm.py:300: error: Incompatible return value type (got "Union[Coroutine[Any, Any, Any], Any]", expected "int")
web3/pm.py:303: error: Incompatible return value type (got "Union[Coroutine[Any, Any, Any], Any]", expected "int")
Found 12 errors in 2 files (checked 176 source files)
```

Reply:
Typing is killing us right now 😆 I think it makes sense to keep them in the child classes. We can re-evaluate if it gets too clunky later.